**Department of Mathematics <br>
College of Arts and Sciences, Howard University <br>
MATH 014 (01) – Introduction to Data Science** <br>
**Project 2 – Part 1** <br>
**Student Name:** Halia Benn <br>
**Student ID:** 004001244 <br>
**Due date: 04-4-2025**

---

## **Introduction** 
### **Project Title:** 2000-2023 Global Environmental Trends 
**Description:** This dataset is a record of Global Warming-related reports from 14 countries between the year 2000 and 2023. This dataset provides the potential for analyzing intricate relationships human impacts, changing conditions of the natural environment, and climate variables, all of which affect the trajectory of the global climate change crisis the entire world is united against. This dataset is helpful for those interested in climate change, environmental impacts, policy, and sustainability.

**Source for reference:** 
https://www.kaggle.com/datasets/adilshamim8/temperature/data

---

### **SECTION I: DATASET OVERVIEW AND STRUCTURE**
1. What is the shape and structure of the dataset? (Use .shape, .info(), .head(), .tail()) <br> 
    * **Show first and last few rows of the data set.**
    * **Describe the general information of the DataFrame.**

In [295]:
# Pandas is a Python library for data manipulation and analysis, helpful for handling tabular, structured data 
import pandas as pd # import as user-defined "pd"
print('Pandas imported as pd.')

# Numpy is a Python library primarily for numarical calculations
import numpy as np  # import as user-defined "np"
print('Numpy imported as np.')

Pandas imported as pd.
Numpy imported as np.


In [297]:
# Load the dataset with the built-in function pd.read_csv, calling the file by name
df = pd.read_csv('temperature.csv') 
print('Dataset "temperature.csv" loaded.')

Dataset "temperature.csv" loaded.


In [299]:
# The built-in function pd.shape gives the dimensions of the dataset (rows, columns)
print('Dimensions of the dataset:\n', df.shape) # print the command

Dimensions of the dataset:
 (1000, 10)


In [301]:
# The dataset is too large to display as a whole, so printing it will display the first and last few rows by default 
print('First and last few rows of the dataset:\n') 
df # display the dataset # tip: avoid using print() with df commands for neater formatting
# Alternatively, the built-in functions df.head() and df.tail() display the first and last 5 rows of the dataset by default

First and last few rows of the dataset:



Year    Country  Avg_Temperature_degC  CO2_Emissions_tons_per_capita  \
0    2006         UK                   8.9                            9.3   
1    2019        USA                  31.0                            4.8   
2    2014     France                  33.9                            2.8   
3    2010  Argentina                   5.9                            1.8   
4    2007    Germany                  26.9                            5.6   
..    ...        ...                   ...                            ...   
995  2019      India                  23.6                            8.0   
996  2000         UK                  21.8                           10.0   
997  2019  Argentina                  23.8                            8.9   
998  2016  Australia                  21.0                           14.9   
999  2011    Germany                  24.1                           17.3   

     Sea_Level_Rise_mm  Rainfall_mm  Population  Renewable_Energy_pct  \
0                  3.1         1441   530911230                  20.4   
1                  4.2         2407   107364344                  49.2   
2                  2.2         1241   441101758                  33.3   
3                  3.2         1892  1069669579                  23.7   
4                  2.4         1743   124079175                  12.5   
..                 ...          ...         ...                   ...   
995                1.2         1365  1358019778                  10.0   
996                2.2         1273   876123161                  14.9   
997                4.7          891  1120533308                  25.9   
998                3.1         1136   380662109                  24.5   
999                2.1         2854   398407112                  41.0   

     Extreme_Weather_Events  Forest_Area_pct  
0                        14             59.8  
1                         8             31.0  
2                         9             35.5  
3                         7             17.7  
4                         4             17.4  
..                      ...              ...  
995                       8             20.2  
996                      14             30.1  
997                      10             46.5  
998                       3             44.5  
999                       3             19.8  

[1000 rows x 10 columns]

In [303]:
# The built-in function df.info displays general info such as data types and null values
print('General information about the dataset:\n')
df.info() # print the command 

General information about the dataset:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 10 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Year                           1000 non-null   int64  
 1   Country                        1000 non-null   object 
 2   Avg_Temperature_degC           1000 non-null   float64
 3   CO2_Emissions_tons_per_capita  1000 non-null   float64
 4   Sea_Level_Rise_mm              1000 non-null   float64
 5   Rainfall_mm                    1000 non-null   int64  
 6   Population                     1000 non-null   int64  
 7   Renewable_Energy_pct           1000 non-null   float64
 8   Extreme_Weather_Events         1000 non-null   int64  
 9   Forest_Area_pct                1000 non-null   float64
dtypes: float64(5), int64(4), object(1)
memory usage: 78.3+ KB


In [305]:
# int64 indicates a numerical entry, while object indicates a categorical enrtry

**2. Identify the variable types (categorical, numerical, datetime, etc.)** 

In [308]:
# The built-in function df.datatypes prints a string of the datatypes for each column  
print('Variable types of the dataset:\n')
df.dtypes # print the function

Variable types of the dataset:



Year                               int64
Country                           object
Avg_Temperature_degC             float64
CO2_Emissions_tons_per_capita    float64
Sea_Level_Rise_mm                float64
Rainfall_mm                        int64
Population                         int64
Renewable_Energy_pct             float64
Extreme_Weather_Events             int64
Forest_Area_pct                  float64
dtype: object

In [310]:
# int64 represents integer-based data (ex. 1,2,3)
# object represents string-based data (ex. one, two, three)
# float represents decimal-based data (ex. 1.0, 2.0, 3.0)

3. Are there any missing values? <br> 
    * How many Missing Values?
    * Briefly describe how you plan to handle them.

In [313]:
# The built-in function df.isnull() identifies each value as a true or false that it is a null value
print('Total of missing values in the dataset:\n')
df.isnull().sum() # use .sum() to print the total # of missing values in each column

Total of missing values in the dataset:



Year                             0
Country                          0
Avg_Temperature_degC             0
CO2_Emissions_tons_per_capita    0
Sea_Level_Rise_mm                0
Rainfall_mm                      0
Population                       0
Renewable_Energy_pct             0
Extreme_Weather_Events           0
Forest_Area_pct                  0
dtype: int64

In [315]:
# there are no null values in the dataset

In [317]:
# Hypothetically if there were null values to handle:
# if missing data is less than 20% of the column, we can relace the null values with the mode of the column
# alternatively, we could replace the values with "Unknown" or drop the column entirely

---

### **SECTION II: DATA CLEANING & INDEXING**
4. Check for and remove duplicates. (Use .duplicated().sum() and .drop_duplicates() if applicable)
5. Provide two examples of DataFrame filtering using logical conditions (e.g., by category, threshold). You can include more!

In [320]:
# before doing anything to alter the data, we will copy it. This is important to make sure we don't ruin the original data file. 
df_new = df.copy()

In [322]:
# .duplicated().sum() is a built in function to calulate the sum of all duplicate data entries.
df_new.duplicated().sum()

0

In [324]:
# Since it returned 0, there are no duplicate counts. If there were duplicates, we could use .drop_duplicates to remove them from the dataset

In [410]:
# .groupby is a built-in function that enables data filtering through grouping it by columns of choice. It can be used to see the intersections of columns

# created a user-defined name for this filter, assign value of the dataset copy grouped by Country, extracting the mean values of CO2 emissions for each country
CO2_by_country=df_new.groupby('Country')['CO2_Emissions_tons_per_capita'].mean().sort_values().reset_index(name='CO2 Emissions (tons/capita)')

# .sort_values() is a built in funtion that orders the data. By default, it has sorted it from low to high 
# .reset_index is a built-in funtion that resets an index label. We can then assign a new name to it 

print('C02 Emissions (tons per capita) by Country:\n')
print(CO2_by_country) # print the filter 

C02 Emissions (tons per capita) by Country:

           Country  CO2 Emissions (tons/capita)
0        Australia                     8.900000
1           Russia                     8.967568
2           Mexico                     9.347273
3   'South Africa'                     9.617808
4          Germany                     9.880328
5           Canada                    10.113433
6        Argentina                    10.447761
7            India                    10.694286
8            China                    10.835821
9           Brazil                    10.838806
10             USA                    10.882192
11           Japan                    10.960317
12          France                    10.966667
13       Indonesia                    11.021333
14              UK                    12.650769


**Interpretation:** <br>
* Australia has the lowest carbon dioxide emissions in tons per capita, while the United Kingdom has the highest. The difference between them is 3.750769 tons per capita.
* Majority of the countries producing the most CO2 emissions are known to be more industry heavy.

**Why it matters:** <br>
* **This helps to put into perspective the countries as CO2 emission producers, in relation to each other**
* The tons per capita metric of measurement accounts for the size of each country. It is the total emissions divided by the population to compare the average amount produced by a single person.
* This standardized mesurement arguably brings into question the kind of culture being fostered towards environmental impact in each country, when considering the daily requirements of a person's livestyle that may affect the global atmosphere.
  - However, I am curious if this removes a level of accountability from large industrial coorporations that produce a level of C02 equivalent to many citizens combined.

In [426]:
# use .groupby to filter the data by multiple columns at once 
tb =df_new.groupby(['Year','Country'])['CO2_Emissions_tons_per_capita'].mean().reset_index(name='CO2 Emissions (tons/capita)') 
# .reset_index is a built-in funtion that resets an index label. We can then assign a new name to it 

print('C02 Emissions of each country in 2000:\n')
tb.head(14) # print the first 14 rows for all 14 countries in the first year documented

C02 Emissions of each country in 2000:



Year         Country  CO2 Emissions (tons/capita)
0   2000  'South Africa'                    13.985714
1   2000       Argentina                     3.900000
2   2000       Australia                    10.466667
3   2000          Brazil                    14.600000
4   2000          Canada                     8.850000
5   2000           China                     6.925000
6   2000          France                    14.600000
7   2000         Germany                     5.950000
8   2000           India                    12.250000
9   2000       Indonesia                    13.942857
10  2000          Mexico                     4.400000
11  2000          Russia                     9.900000
12  2000              UK                    10.987500
13  2000             USA                    13.040000

In [428]:
print('CO2 Emissions of each country in 2023:\n')
tb.tail(15) # print the last 14 rows for all 14 countries in the last year documented 

CO2 Emissions of each country in 2023:



Year         Country  CO2 Emissions (tons/capita)
325  2023  'South Africa'                    15.000000
326  2023       Argentina                    12.283333
327  2023       Australia                     9.200000
328  2023          Brazil                    13.100000
329  2023          Canada                     7.900000
330  2023           China                     9.333333
331  2023          France                    11.625000
332  2023         Germany                    10.900000
333  2023           India                    17.650000
334  2023       Indonesia                    12.900000
335  2023           Japan                     4.800000
336  2023          Mexico                     9.750000
337  2023          Russia                    11.400000
338  2023              UK                    14.300000
339  2023             USA                    11.242857

**Interpretation:** <br>
* Between 2000 and 2023, some countries have increased in C02 emissions, such as Argentina that had a dramatic increase from 3.9 tons/capita to 12.28 tons per capita
* Others have seen decreases, such as Russia, USA, and Frace 

**Why it matters:** <br>
* While it is difficult to view the entire dataframe, at least comparing the first and last years available for each country **gives insight into how CO2 emissions have progressed from each of them, over the span of 23 years**. However, this is not to assume a linear trajectory of each change.
* I am curious to look into the historical occurences such as policy change or the success of particular industries that have affected the levels of CO2 emissions from a country. For example, the environmental policies of the Clinton administration in 2000 versus the Biden administration in 2023, in case their leadership has anything to do with the levels of C02 produced during their terms.
* The decreases of countries in the UN (United Nations) could be due to pacts such as the Paris Agreement established to mitigate global warming 

---

### **SECTION III: DESCRIPTIVE STATISTICS**
6. Use .describe() on numerical columns. Share at least one interesting insight.<br>
7. Use .value_counts() to analyze at least two categorical columns.<br>
8. Use .groupby() to compute summary statistics (e.g., mean or count) across categories.<br>
Note: If you find any additional interesting filtering/grouping do not forget to include that!

In [462]:
print('Summary Statistics for Extreme Weather Events:\n')
df_new['Extreme_Weather_Events'].describe()

Summary Statistics for Extreme Weather Events:



count    1000.000000
mean        7.291000
std         4.422655
min         0.000000
25%         3.000000
50%         8.000000
75%        11.000000
max        14.000000
Name: Extreme_Weather_Events, dtype: float64

In [460]:
print('Summary Statistics for Forest Area:\n')
df_new['Forest_Area_pct'].describe()

Summary Statistics for Forest Area:



count    1000.000000
mean       40.572000
std        17.398998
min        10.100000
25%        25.600000
50%        41.150000
75%        55.800000
max        70.000000
Name: Forest_Area_pct, dtype: float64

**Notable Insights:**
* There was a country with 70% of its land occupied by forests at some point in time, and one with 10%.
* When viewed as an indicator of ecological health, this is a vast difference. It could be that the 10% naturally had less forests to begin with, or that areas were deforested for industrial purposes, for example.

In [458]:
print('Summary Statistics for Renewable Energy:\n')
df_new['Renewable_Energy_pct'].describe()

Summary Statistics for Renewable Energy:



count    1000.000000
mean       27.300500
std        12.970808
min         5.100000
25%        16.100000
50%        27.150000
75%        38.925000
max        50.000000
Name: Renewable_Energy_pct, dtype: float64

**Notable Insights:**
* The highest amount of renewable energy in a country was 50% at some point in time.
* If the record to beat was half of all energy in a country being renewable, what is the goal or record that nations are pushing for today? Do we nave new technology to support this goal? 

In [456]:
print('Summary Statistics for Population:\n')
df_new['Population'].describe()

Summary Statistics for Population:



count    1.000000e+03
mean     7.053830e+08
std      4.093910e+08
min      3.660891e+06
25%      3.436242e+08
50%      7.131166e+08
75%      1.073868e+09
max      1.397016e+09
Name: Population, dtype: float64

In [464]:
print('Summary Statistics for Extreme Weather Events:\n')
df_new['Year'].describe()

Summary Statistics for Extreme Weather Events:



count    1000.000000
mean     2011.432000
std         7.147199
min      2000.000000
25%      2005.000000
50%      2012.000000
75%      2018.000000
max      2023.000000
Name: Year, dtype: float64

In [482]:
# .value_counts is a built-in function that returns the frequency of each uniqe category in a column
print('Unique value counts for Country:')
df_new['Country'].value_counts()

Unique value counts for Country:


Country
Indonesia         75
Russia            74
USA               73
'South Africa'    73
India             70
Argentina         67
China             67
Brazil            67
Canada            67
France            66
UK                65
Japan             63
Germany           61
Australia         57
Mexico            55
Name: count, dtype: int64

**Interpretation:**
* Indonesia is represented the most in the dataset (75 times)
* Mexico is represented the least in the dataset (55 times)

**Why it matters:**
* It is uncertain why each country was not accounted for in all 23 years
* This is important to note when considering statistics involving the country column, because there is not an equal representation of each country 

In [349]:
# Since 'Country' is the only categorical column, I will create another to analyze. 
# This new column will be based on the summary statistics of the 'Population column'

In [478]:
def classify_population(population): # define a user-defined function with the def Python feature 
    if population <= 3.436242e+08: # use loop-based logic to establish a minimum from the Population column
        return 'Small' # if the number is less than or equal to the minimum, it will be a small population 
    elif population <= 1.073868e+09: # use elif function to create additional if statements for the loop
        return 'Medium' # using the median of the Population column
    else:
        return 'Large' # all else, meaning above the interquartile range (IQR) of the Population column, will be large

# create a new column that organizes existing data from the Population column with this new function
df_new['Population Size'] = df_new['Population'].apply(classify_population) # apply user-defined function to the Population column
df_new # print for reference

Year    Country  Avg_Temperature_degC  CO2_Emissions_tons_per_capita  \
0    2006         UK                   8.9                            9.3   
1    2019        USA                  31.0                            4.8   
2    2014     France                  33.9                            2.8   
3    2010  Argentina                   5.9                            1.8   
4    2007    Germany                  26.9                            5.6   
..    ...        ...                   ...                            ...   
995  2019      India                  23.6                            8.0   
996  2000         UK                  21.8                           10.0   
997  2019  Argentina                  23.8                            8.9   
998  2016  Australia                  21.0                           14.9   
999  2011    Germany                  24.1                           17.3   

     Sea_Level_Rise_mm  Rainfall_mm  Population  Renewable_Energy_pct  \
0                  3.1         1441   530911230                  20.4   
1                  4.2         2407   107364344                  49.2   
2                  2.2         1241   441101758                  33.3   
3                  3.2         1892  1069669579                  23.7   
4                  2.4         1743   124079175                  12.5   
..                 ...          ...         ...                   ...   
995                1.2         1365  1358019778                  10.0   
996                2.2         1273   876123161                  14.9   
997                4.7          891  1120533308                  25.9   
998                3.1         1136   380662109                  24.5   
999                2.1         2854   398407112                  41.0   

     Extreme_Weather_Events  Forest_Area_pct Population Size  \
0                        14             59.8          Medium   
1                         8             31.0           Small   
2                         9             35.5          Medium   
3                         7             17.7          Medium   
4                         4             17.4           Small   
..                      ...              ...             ...   
995                       8             20.2           Large   
996                      14             30.1          Medium   
997                      10             46.5           Large   
998                       3             44.5          Medium   
999                       3             19.8          Medium   

     Total C02 Emissions (tons)  
0                  4.937474e+09  
1                  5.153489e+08  
2                  1.235085e+09  
3                  1.925405e+09  
4                  6.948434e+08  
..                          ...  
995                1.086416e+10  
996                8.761232e+09  
997                9.972746e+09  
998                5.671865e+09  
999                6.892443e+09  

[1000 rows x 12 columns]

**What it does:** 
* Classifies each country by population size

**Why it helps:**
* Makes populations of each country easy to compare in relation to each other (based on the summary statistics of the all populations in the dataset)
* Because the Populations column consists of such large numbers without commas, it is difficult to make comparisons

In [480]:
print('Unique value counts for Population Size:\n')
df_new['Population Size'].value_counts()

Unique value counts for Population Size:



Population Size
Medium    500
Small     250
Large     250
Name: count, dtype: int64

**Interpretation:**
* The majority (half) of the data is from countries with population sizes average/ medium in comparison to all populations in the dataset
* One fourth of the data is from smaller polulations (relative to the dataset)
* One fourth of the data is from larger populations (relative to the dataset)
* **take into account that there are only 14 countries, and the unique values of the Country column shows that they are not evenly accounted for in each year**

**Why it matters:**
* **It shows the proportion of population sizes in relation to each other**
* the size of the country matters in relation to other columns such as CO2 emissions and Forest Area

In [440]:
# create a filter to see the average C02 emissions of each population size 
co2_by_size = df_new.groupby('Population Size')['CO2_Emissions_tons_per_capita'].mean().reset_index(name='CO2 Emissions (tons/capita)')

print('CO2 Emissions by Population Size:\n')
co2_by_size 

CO2 Emissions by Population Size:



Population Size  CO2 Emissions (tons/capita)
0           Large                      10.3924
1          Medium                      10.3594
2           Small                      10.5920

**Interpretation:**
* Nearly an even distribution of tons per capita across all population sizes
* countries with smaller population sizes in relation to the dataset produce the most carbon dioxide emissions by a small amount
  - however, this small amount is significant in that it adds up when multiplied by the population 

**Why it matters:**
* shows that **population size is not a major defining factor of the average C02 emissions per capita** produced by countries

In [379]:
# creating a new column to calculate the total tons by multiplying the Population column and the C02 emissions (tons/capita) column
df_new['Total C02 Emissions (tons)'] = df_new['CO2_Emissions_tons_per_capita']* df_new['Population']
df_new # print the dataframe for reference 

Year    Country  Avg_Temperature_degC  CO2_Emissions_tons_per_capita  \
0    2006         UK                   8.9                            9.3   
1    2019        USA                  31.0                            4.8   
2    2014     France                  33.9                            2.8   
3    2010  Argentina                   5.9                            1.8   
4    2007    Germany                  26.9                            5.6   
..    ...        ...                   ...                            ...   
995  2019      India                  23.6                            8.0   
996  2000         UK                  21.8                           10.0   
997  2019  Argentina                  23.8                            8.9   
998  2016  Australia                  21.0                           14.9   
999  2011    Germany                  24.1                           17.3   

     Sea_Level_Rise_mm  Rainfall_mm  Population  Renewable_Energy_pct  \
0                  3.1         1441   530911230                  20.4   
1                  4.2         2407   107364344                  49.2   
2                  2.2         1241   441101758                  33.3   
3                  3.2         1892  1069669579                  23.7   
4                  2.4         1743   124079175                  12.5   
..                 ...          ...         ...                   ...   
995                1.2         1365  1358019778                  10.0   
996                2.2         1273   876123161                  14.9   
997                4.7          891  1120533308                  25.9   
998                3.1         1136   380662109                  24.5   
999                2.1         2854   398407112                  41.0   

     Extreme_Weather_Events  Forest_Area_pct Population Size  \
0                        14             59.8          Medium   
1                         8             31.0           Small   
2                         9             35.5          Medium   
3                         7             17.7          Medium   
4                         4             17.4           Small   
..                      ...              ...             ...   
995                       8             20.2           Large   
996                      14             30.1          Medium   
997                      10             46.5           Large   
998                       3             44.5          Medium   
999                       3             19.8          Medium   

     Total C02 Emissions (tons)  
0                  4.937474e+09  
1                  5.153489e+08  
2                  1.235085e+09  
3                  1.925405e+09  
4                  6.948434e+08  
..                          ...  
995                1.086416e+10  
996                8.761232e+09  
997                9.972746e+09  
998                5.671865e+09  
999                6.892443e+09  

[1000 rows x 12 columns]

**What it does:**
* Shows the total C02 emissions of each country 

**Why it helps:** 
* Just for reference of the total amount of C02 emissions in tons for each country, in comparison to the standardized metric by person
* More information to contextualize the available data 

In [442]:
# use the total C02 emissions column to see the average total of each population size 
total_CO2_by_size = df_new.groupby('Population Size')['Total C02 Emissions (tons)'].mean().reset_index(name = 'Total C02 Emissions (tons)')

print('Total C02 Emissions by Population Size:\n')
total_CO2_by_size 

Total C02 Emissions by Population Size:



Population Size  Total C02 Emissions (tons)
0           Large                1.279933e+10
1          Medium                7.404640e+09
2           Small                1.914016e+09

**Interpretation:** 
* Larger countries produced an average total of nearly 12.8 billion tons of C02 emissions
* Medium countries produced an average total of about 7.4 billion tons of C02 emissions
* Smaller countries produced an average total of about 1.9 billion tons of CO2 emissions

**Why it matters:**
* Demonstrates the difficulty in comparing CO2 emissions meaningfully without a standardized metric of measurement to account for the population size

---

### **SECTION IV: OBSERVATIONS & QUESTIONS**
9. Highlight 1–2 key takeaways or patterns from your descriptive analysis.
10. Suggest 2 questions you would explore further in Part 2 using visualization.


**Key takeaways from this descriptive analysis:**
* It was a bit difficult finding relationships between the columns, except for population, year, country, and C02 emissions. I learned that population size does not necessarily impact the average C02 emissions produced by the 14 countries, surprisingly
* There seems to be more importance in each country having a unique unity of factors that define its relationship to the global climate 

**Questions for further analysis using visualization:**
* How can the columns that were not analyzed (such as sea level rise, temperature, extreme weather events) be meaningfully engaged to create a more effective understanding of the changing natural conditions unique to each of the 14 countries? Could I instead find more commonality than difference between them all?
* Could I illustrate the relationship between human impacts (such as industrial activity causing C02 emissions or population) and natural environment changes (temperature or forest area) through visual representation?

---

### **Conclusion:**
The 14 countries included in the dataset all seem to have an impact on the environment relative to each other. From a descriptive analysis mainly focused on C02 emissions and population size, each grouping was relatively in balance with the others. This is likely attributed to the use of broad groupings and mean values to describe them. This was purely based on a need to cross-analyze between columns with the available information given. It was difficult to find relationships of direct or inverse variation between the columns because there were few with commonly known correlation between them. This could also require some expertise on environmental impact studies. However, visualization may help in viewing the unique environmenal situations of each country in relation to each other. 